pkg/planner/core/plan_cost_ver2.go

```diff
@@ -141,6 +141,10 @@ func (p *PhysicalTableScan) GetPlanCostVer2(taskType property.TaskType, option *
 	// give TiFlash a start-up cost to let the optimizer prefers to use TiKV to process small table scans.
 	if p.StoreType == kv.TiFlash {
 		p.PlanCostVer2 = costusage.SumCostVer2(p.PlanCostVer2, scanCostVer2(option, 10000, rowSize, scanFactor))
+	} else if !p.isChildOfIndexLookUp && (p.SCtx().GetSessionVars().GetAllowPreferRangeScan() && (p.tblColHists.Pseudo || p.tblColHists.RealtimeCount < 1)) || p.tblColHists.ModifyCount > p.tblColHists.RealtimeCount {
```

Comment:
The function GetPlanCostVer2 is quite complicated, hard to maintain. And There some some places we can further optimize. Such as remove magic numbers, something like this would be more clear.
```go
const (
    MinRowSize             = 2.0
    TiFlashStartupRowPenalty = 10000
    MaxPenaltyRowCount     = 1000
)

// GetPlanCostVer2 returns the plan-cost of this sub-plan, which is:
// plan-cost = rows * log2(row-size) * scan-factor
// log2(row-size) is from experiments.
func (p *PhysicalTableScan) GetPlanCostVer2(taskType property.TaskType, option *optimizetrace.PlanCostOption) (costusage.CostVer2, error) {
    if p.PlanCostInit && !hasCostFlag(option.CostFlag, costusage.CostFlagRecalculate) {
        return p.PlanCostVer2, nil
    }

    rows := getCardinality(p, option.CostFlag)
    
    var columns []Column
    if p.StoreType == kv.TiKV {
        columns = p.tblCols
    } else { // TiFlash
        columns = p.schema.Columns
    }
    rowSize := getAvgRowSize(p.StatsInfo(), columns)
    // Ensure rowSize has a reasonable minimum value to avoid underestimation
    rowSize = math.Max(rowSize, MinRowSize)
    
    scanFactor := getTaskScanFactorVer2(p, p.StoreType, taskType)
    p.PlanCostVer2 = scanCostVer2(option, rows, rowSize, scanFactor)

    // Apply TiFlash startup cost to prefer TiKV for small table scans
    if p.StoreType == kv.TiFlash {
        p.PlanCostVer2 = costusage.SumCostVer2(p.PlanCostVer2, scanCostVer2(option, TiFlashStartupRowPenalty, rowSize, scanFactor))
    } else {
        sessionVars := p.SCtx().GetSessionVars()
        allowPreferRangeScan := sessionVars.GetAllowPreferRangeScan()
        tblColHists := p.tblColHists

        preferRangeScanCondition := allowPreferRangeScan && (tblColHists.Pseudo || tblColHists.RealtimeCount < 1)
        hasHighModifyCount := tblColHists.ModifyCount > tblColHists.RealtimeCount
        shouldApplyPenalty := !p.isChildOfIndexLookUp && (preferRangeScanCondition || hasHighModifyCount)

        if shouldApplyPenalty {
            newRowCount := math.Min(MaxPenaltyRowCount, math.Max(float64(tblColHists.ModifyCount), float64(tblColHists.RealtimeCount)))
            p.PlanCostVer2 = costusage.SumCostVer2(p.PlanCostVer2, scanCostVer2(option, newRowCount, rowSize, scanFactor))
        }
    }

    p.PlanCostInit = true
    return p.PlanCostVer2, nil
}

```